lib/active_merchant/billing/gateways/payu_latam.rb

```diff
@@ -0,0 +1,400 @@
+          post[:transaction][:payer] = payer
+        end
+      end
+
+      def add_extra_parameters(post, options)
+        extra_parameters = {}
+        extra_parameters[:INSTALLMENTS_NUMBER] = options[:installments_number] || 1
+        post[:transaction][:extraParameters] = extra_parameters
+      end
+
+      def add_reference(post, authorization)
+        order_id, transaction_id = split_authorization(authorization)
+        order = {}
+        order[:id] = order_id
+        post[:transaction][:order] = order
+        post[:transaction][:parentTransactionId] = transaction_id
+        post[:transaction][:reason] = 'n/a'
+      end
+
+      def add_payment_method_to_be_tokenized(post, payment_method)
+        credit_card_token = {}
+        credit_card_token[:payerId] = generate_unique_id
+        credit_card_token[:name] = payment_method.name.strip
+        credit_card_token[:identificationNumber] = generate_unique_id
+        credit_card_token[:paymentMethod] = BRAND_MAP[payment_method.brand.to_s]
+        credit_card_token[:number] = payment_method.number
+        credit_card_token[:expirationDate] = format(payment_method.year, :four_digits).to_s + '/' + format(payment_method.month, :two_digits).to_s
+        credit_card_token[:securityCode] = payment_method.verification_value
+        post[:creditCardToken] = credit_card_token
+      end
+
+      def commit(action, params)
+        begin
+          raw_response = ssl_post(url, post_data(params), headers)
+          response = parse(raw_response)
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response_error(raw_response)
+        rescue JSON::ParserError
+          unparsable_response(raw_response)
+        else
+          success = success_from(action, response)
+          Response.new(
+            success,
+            message_from(action, success, response),
+            response,
+            authorization: success ? authorization_from(action, response) : nil,
+            error_code: success ? nil : error_from(action, response),
+            test: test?
+          )
+        end
+      end
+
+      def headers
+        {
+          "Content-Type"  => "application/json",
+          "Accept"  => "application/json"
+        }
+      end
+
+      def post_data(params)
+        params.to_json
+      end
+
+      def url
+        test? ? test_url : live_url
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def success_from(action, response)
+        case action
+        when 'store'
+          response["code"] == "SUCCESS" && response["creditCardToken"] && response["creditCardToken"]["creditCardTokenId"].present?
+        when 'verify_credentials'
+          response["code"] == "SUCCESS"
+        else
+          response["code"] == "SUCCESS" && response["transactionResponse"] && (response["transactionResponse"]["state"] == "APPROVED")
+        end
+      end
+
+      def message_from(action, success, response)
```

Comment:
Any good ideas how we might refactor this a bit?  Perhaps extracting parts into other methods?